src/Sulu/Bundle/AdminBundle/Admin/View/FormOverlayListViewBuilderInterface.php

```diff
@@ -97,4 +97,6 @@ public function addMetadataRequestParameters(array $metadataRequestParameters):
      * @param Badge[] $badges
      */
     public function addTabBadges(array $badges): self;
+
+    public function addErrorCodeMessages(array $errorCodeMessages): self;
```

Comment:
I must say I don't lik it this way to configure this on the view level. The error `codes` are designed to be unique and every package had its own [number](https://github.com/sulu/docs/tree/master/detail-specification). Not sure if I would implement this things this way. I think still the whole correct translated messages should directly come from the error response and no mapping between an error code or other things should be needed to be configured somewhere. And if it should we maybe should configure it at a global place and not for every view again. As documented in the issue I really would prefer that we use a well formated error format like its done in the RFC and and not had additional mapping on the view here.

/cc @sulu/core-team

Reply:
The idea behind this approach is that we dont need to use a specific format for the `code`. It allows to display messages for numerical codes of the existing API, but we can also use string codes such as `blocked-filetype` in newly implemented APIs.

Another idea is to make it possible to display different error messages for the same code based on the current view. For example, this makes it easily possible to **display an error message based on the HTTP status code of the response**. Lets be honest - most developers will just throw a HTTP status code to indicate an error. 

> The error codes are designed to be unique and every package had its own number.
This does not scale very well 🙈  Who decides which third party bundle uses which code range?

> maybe a global mapping between error code and error message would be better.
As said above, a global mapping does not scale very well. Also it prevents to display a error message for a specific status code. What advantages do you see with that?

> I think still the whole correct translated messages should directly come from the error response and no mapping between an error code or other things should be needed to be configured somewhere.
That wont be easy for unlocalized APIs. Do you know an API that returns translated messages? I would prefer to return local independent key such as `blocked-filetype` in newly implemented APIs. 

> As documented in the issue I really would prefer that we use a well formated error format like its done in the RFC and and not had additional mapping on the view here.
We should still implement a well structured error format that allows to display an error for a specific field in a later PR.  We could also add a `errorMessage` property that will be displayed directly (without any mapping and without translating in in the frontend) in a later PR 🙂

Reply:
We cannot change the existing APIs and we should really implement a solution for this. It is quite embarrassing that Sulu still does not allow to display a custom error message at all. 

This solution **does not require big changes or BC breaks** and **works with the existing API**. Furthermore it is flexible enough to be used inside of projects or bundles and it does not prevent to implement a better solution for displaying field-specific errors in the future.

Reply:
> but we can also use string codes such as blocked-filetype in newly implemented APIs.

Definitly we should not go on "string" codes. I also think string codes would be the false way. I think the API should return a translated error message for the user system language. So also API only users know really what the error was.

> Returning a localized message is kind of strange for unlocalized APIs.  I would prefer to return local independent key such as blocked-filetype in newly implemented APIs.

Adding a key is maybe a good idea to add logic on it but as we have the code for it I would not add something like a key.

> Returning a localized message is kind of strange for unlocalized APIs.

Error messages are not returned in the "content" language of the API they are returned in the "system" languages. SO I don't see here a problem with unlocalized APIs.

> Do you know an API that returns translated messages?

The most API i did work with always had both. Some kind of error code or error string key and a translated message, so you never needed to lookup errors in some documentation. 

> This solution does not require big changes or BC breaks, works with the existing API 

But we are providing then different ways to achieve the same thing. Which I think is not a great idea. And what we do in future when there is a error message returned which has the priority the mapping in the view or the error message returned by the API?

> BC breaks

I don't think that adding properties to the error response is a BC Break. So I did never understand that argument from daniel as we did add properties on all apis without any versioning and problems. So the current state is:

```json
{
     "code": "0",
     "message": "Internal Server Error"
}
```

And then add here the rfc7807 message parameters would be no bc break:

```json
{
     "code": "0",
     "message": "...",
     "type": "..",
     "title": ".."
}
```

I just wanted to mention my fears about implementing it this way. Also I'm afraid that we will never implement it to provide nice error messages over our APIs which I think is required for a good Developer Experience. I hope you understand my point of view on this topic. I really appreciate that you are working on this and if the others don't share my fears I'm okay with the solution.

Reply:
> Definitly we should not go on "string" codes. I also think string codes would be the false way. I think the API should return a translated error message for the user system language.

I think string codes are nicer than number codes because the make the code easier to understand without the need of constants, we can still add a translated message in another key. This could look like this:
```
{
    "code": "blocked-filetype"
    "message": "The filetype of the uploaded file is configured to be blocked."
}
```

> But we are providing then different ways to achieve the same thing. Which I think is not a great idea.

That is right, there would be two possibilities:

- Specify the message for an error code or HTTP status code in a mapping
- Specify the error message that should be displayed directly

We dont need to use the first possibility in the core, but I would still like to implement it because I think most projects will just use HTTP status codes for indicating errors.

> And what we do in future when there is a error message returned which has the priority the mapping in the view or the error message returned by the API?

I would use the following priority:

1. Translated error message in the response (eg `message` property)
2. Mapped error message for the `code` property of the response
3. Mapped error message for the HTTP status code of the response

It would be okay for me to get rid of the second type. But as already said, I think we should allow to define a mapping for HTTP status codes 🙂 

> Also I'm afraid that we will never implement it to provide nice error messages over our APIs which I think is required for a good Developer Experience.

I am afraid of that too. Do you see an easy solution to implement this for the existing APIs? 

If we find an easy way for returning a translated error message in the cases where we already return a code now, I would really like to implement that.

> I really appreciate that you are working on this and if the others don't share my fears I'm okay with the solution.

Thats nice to hear and I appreciatee the feedback! 🙂  
I am sorry if my wording is a bit harsh, but I am really annoyed about this topic and that we still have not made any progress about it. 

**I want to solve this, at least for simple cases, in the near future**. I think this PR is not perfect, but it would solve things for most cases. If we find a way to return additional translated error messages, it would be even better. But I think I need your help for that 🙂